File: AdventOfCode/2024/03.go
package main

import (
  "fmt"
  "os"
  "regexp"
  "strconv"
  "strings"
)

// Read the input data from "input.txt" file. Parse the data into a string of
// the "memory dump" and return it.
func readPuzzleInput() string {
  file, err := os.ReadFile("../input.txt")
  if err != nil {
    panic(err)
  }

  memory := string(file)
  return memory
}

// Find all of the digits inside of a "mul()", ie mul(1,2) would give us [1,2].
// Return a list of all matches
func getMatches(memory string) [][]string {
  re := regexp.MustCompile(`mul\((\d+),(\d+)\)`)
  return re.FindAllStringSubmatch(memory, -1)
}

// Get all of the "mul()" digits from "getMatches()" function, multiply each
// pair of digits and then add their products together. Return the sum of
// products.
func partOne(memory string) int {
  sum := 0
  matches := getMatches(memory)
  for _, match := range matches {
    a, err1 := strconv.Atoi(match[1])
    b, err2 := strconv.Atoi(match[2])
    if err1 != nil || err2 != nil {
      fmt.Println("Error converting to int:", err1, err2)
      continue
    }
    sum += a * b
  }
  return sum
}

// In part 2 of the puzzle we only sum the values of our digits if they follow
// a "do()" function in the corrupted memory dump.
func partTwo(memory string, enabled bool) int {
  if enabled {
    if strings.Contains(memory, "don't()") {
      i := strings.Index(memory, "don't")
      return partOne(memory[:i]) + partTwo(memory[i:], false)
    } else {
      return partOne(memory)
    }
  } else {
    if strings.Contains(memory, "do()") {
      i := strings.Index(memory, "do()")
      return partTwo(memory[i:], true)
    } else {
      return 0
    }
  }
}

// Orchestrate the program flow by calling the functions to print out the
// solved Part 1 and Part 2 functions of the puzzle.
func main() {
  fmt.Println("Part 1:", partOne(readPuzzleInput()))        // 174960292
  fmt.Println("Part 2:", partTwo(readPuzzleInput(), true))  // 56275602
}

File: AdventOfCode/2024/03.py
#!/usr/bin/env python

import re


def read_puzzle_input() -> str:
    """
    Read the file and create a big string of the data, or the corrupted memory
    per the puzzle so the parts can do their magic with it.

    :return: A big ol' string of corrupted memory data
    :rtype: str
    """
    with open("03.in", "r") as file:
        memory = file.read()
    return memory


def get_matches(memory: str) -> list:
    """
    Find all of the digits inside of a "mul()", ie mul(1,2) would give us
    [1,2].

    :param str memory: The memory dump from the puzzle input
    :return: List of all the "mul()" digits
    :rtype: list
    """
    matches = re.findall(r'mul\((\d+),(\d+)\)', memory, re.MULTILINE)
    return matches


def part_one(memory: str) -> int:
    """
    Get all of the "mul()" digits from "get_matches()" function, multiply each
    pair of digits and then add their products together.

    :param str memory: The memory dump from the puzzle input
    :return: The sum, aka the answer of part 1
    :rtype: int
    """
    results = [int(a) * int(b) for a, b in get_matches(memory)]
    return sum(results)


def part_two(memory: str, enabled: bool) -> int:
    """
    In part 2 of the puzzle we only sum the values of our digits if they follow
    a "do()" function in the corrupted memory dump. Not gonna lie, I needed
    Reddit's help with this one.

    :param str memory: The memory dump from the puzzle input
    :param bool enabled: A logic helper to separate the do's & the dont's math
    :return: The sum, aka the answer of part 2
    :rtype: int
    """
    if enabled:
        if "don't()" in memory:
            i = memory.index("don't")
            return part_one(memory[:i]) + part_two(memory[i:], False)
        else:
            return part_one(memory)
    else:
        if "do()" in memory:
            i = memory.index("do()")
            return part_two(memory[i:], True)
        else:
            return 0


if __name__ == "__main__":
    print("Part 1:", part_one(read_puzzle_input()))        # 174960292
    print("Part 2:", part_two(read_puzzle_input(), True))  # 56275602
